### PR TITLE
[SPARK-33867][SQL] Instant and LocalDate values aren't handled when generating SQL queries

### PR DESCRIPTION
**Pull request by cristichircu**
### What changes were proposed in this pull request?

When generating SQL queries only the old date time API types are handled for values in org.apache.spark.sql.jdbc.JdbcDialect#compileValue. If the new API is used (spark.sql.datetime.java8API.enabled=true) Instant and LocalDate values are not quoted and errors are thrown. The change proposed is to handle Instant and LocalDate values the same way that Timestamp and Date are.

### Why are the changes needed?

In the current state if an Instant is used in a filter, an exception will be thrown.
Ex (dataset was read from PostgreSQL): dataset.filter(current_timestamp().gt(col(VALID_FROM)))
Stacktrace (the T11 is from an instant formatted like yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'): 
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "T11"Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "T11"  Position: 285 at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2103) at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1836) at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:257) at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:512) at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:388) at org.postgresql.jdbc2.AbstractJdbc2Statement.executeQuery(AbstractJdbc2Statement.java:273) at org.apache.spark.sql.execution.datasources.jdbc.JDBCRDD.compute(JDBCRDD.scala:304) at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349) at org.apache.spark.rdd.RDD.iterator(RDD.scala:313) at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52) at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349) at org.apache.spark.rdd.RDD.iterator(RDD.scala:313) at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52) at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349) at org.apache.spark.rdd.RDD.iterator(RDD.scala:313) at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59) at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99) at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52) at org.apache.spark.scheduler.Task.run(Task.scala:127) at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446) at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377) at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.base/java.lang.Thread.run(Thread.java:834)

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Test added

### CODE REVIEW

**Comment by cristichircu** [comment]
@MaxGekk Can you please have a look at this and let me know if there's anything else I can do? Thanks!

**Review by MaxGekk — COMMENTED** [review]
@cristichircu Thanks for the ping. There are integration tests for PostgreSQL in https://github.com/apache/spark/blob/a1a3d5cb02e380156eab320bf6cf512c01b11284/external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala#L38 . Could you add a test there, please.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -175,7 +176,9 @@ abstract class JdbcDialect extends Serializable with Logging{
   def compileValue(value: Any): Any = value match {
     case stringValue: String => s"'${escapeSql(stringValue)}'"
     case timestampValue: Timestamp => "'" + timestampValue + "'"
+    case timestampValue: Instant => "'" + timestampValue + "'"
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:179]: I would guess that the default `Instant` formatting by `toString()` might be not accepted by dbms. Could you take a look at `FractionTimestampFormatter`: https://github.com/apache/spark/blob/3a299aa6480ac22501512cd0310d31a441d7dfdc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala#L121

Comment (cristichircu) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala]: I added the TimestampFormatter. I saw there's a DateFormatter as well. Should I use it for LocalDate or is toString() ok?

**Comment by cristichircu** [comment]
Thanks for pointing me to PostgresIntegrationSuite. Unfortunately I wasn't able to run it on my Windows machine yet. I'll try to run it on a Unix VM or something tomorrow but I expect it will take some time to setup everything since I'm new to Scala and Spark. Any helpful documentation on running the docker integration tests would be appreciated. Thank you!

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @maropu too FYI

**Comment by cristichircu** [comment]
I did the changes mentioned this far.
@MaxGekk, @maropu do you guys have any other concerns?

Any chances this gets in 3.1.1? Anything I can do to help make that happen? 
Thanks!

**Comment by cristichircu** [comment]
This might seem stupid but I'm not sure about the state of the pull request. I see that "All checks have passed" but there's also this failed unit test [here](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134025/testReport/) (that seems unrelated since it probably doesn't used the java 8 date time API).
Please let me know if I need to look into it. Thank you!

**Comment by cristichircu** [comment]
@MaxGekk @maropu anything else I can help with to get this approved?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -1029,6 +1030,19 @@ class JDBCSuite extends QueryTest
       === java.sql.Timestamp.valueOf("2002-02-20 11:22:33.543543"))
   }
 
+  test("Test DataFrame.where for LocalDate and Instant") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:1033]: nit: `test("SPARK-33867: Test DataFrame.where for LocalDate and Instant") {`

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 3]
```diff
@@ -63,6 +66,10 @@ case class JdbcType(databaseTypeDefinition : String, jdbcNullType : Int)
  */
 @DeveloperApi
 abstract class JdbcDialect extends Serializable with Logging{
+
+  private lazy val zoneId = DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone)
+  private lazy val timestampFormatter = TimestampFormatter.getFractionFormatter(zoneId)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:71]: Could you move this inside `compileValue`? Since the dialect object is a singleton, we cannot use `lazy val` here, I think.

**Comment by cristichircu** [comment]
@maropu Thank you for your comments. I made the changes you requested. Please let me know if there's anything else I can help with.
@MaxGekk Do you have anything else to add?

**Comment by cristichircu** [comment]
@MaxGekk @maropu Please let me know if there's anything else I need to do to get this moving. I'm not sure what the process is but it seems like this is stuck and I'm not sure why.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 4]
```diff
@@ -175,7 +179,12 @@ abstract class JdbcDialect extends Serializable with Logging{
   def compileValue(value: Any): Any = value match {
     case stringValue: String => s"'${escapeSql(stringValue)}'"
     case timestampValue: Timestamp => "'" + timestampValue + "'"
+    case timestampValue: Instant =>
+      val timestampFormatter = TimestampFormatter.getFractionFormatter(
+        DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone))
+      "'" + timestampFormatter.format(timestampValue) + "'"
     case dateValue: Date => "'" + dateValue + "'"
+    case dateValue: LocalDate => "'" + dateValue + "'"
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:187]: @MaxGekk, is it the correct way to use the string representation of date/timestamps?

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:187]: Better to use the same as at https://github.com/apache/spark/blob/3a299aa6480ac22501512cd0310d31a441d7dfdc/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala#L200-L206

Comment (cristichircu) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala]: Thanks for pointing that out. Done

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 5]
```diff
@@ -63,6 +66,7 @@ case class JdbcType(databaseTypeDefinition : String, jdbcNullType : Int)
  */
 @DeveloperApi
 abstract class JdbcDialect extends Serializable with Logging{
+
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:69]: Let's remove unrelated newline change.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 6]
```diff
@@ -175,7 +179,14 @@ abstract class JdbcDialect extends Serializable with Logging{
   def compileValue(value: Any): Any = value match {
     case stringValue: String => s"'${escapeSql(stringValue)}'"
     case timestampValue: Timestamp => "'" + timestampValue + "'"
+    case timestampValue: Instant =>
+      val timestampFormatter = TimestampFormatter.getFractionFormatter(
+        DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone))
+      "'" + timestampFormatter.format(timestampValue) + "'"
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:185]: no big deal but I would use `s"'${timestampFormatter.format(timestampValue)}'"` string interpolation

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cristichircu** [comment]
Thanks you for pointing that out, @HyukjinKwon! Done

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master/branch-3.1! Could you backport it into branch-3.0? @cristichircu

**Comment by cristichircu** [comment]
> Thanks! Merged to master/branch-3.1! Could you backport it into branch-3.0? @cristichircu

Done. Opened [pull request](https://github.com/apache/spark/pull/31381) Thanks!
